### ISSUE TITLE
Issues with Array#union and Array#intersect

### ISSUE THREAD

**Issue by andrewplummer** [andrewplummer/Sugar#120]
Moving this discussion over from twitter raised by @ermouth, original example is here: http://ermouth.com/fastArray/

First I want to say thank you for raising this. There are a few different issues here and I want to address them separately. First, here is the detailed result of performing a union operation on your test page:

```
// Array 1
[
    { eccbc87e4b5ce2fe28308fd9f2a7baf3: 3 },
    6,
    3,
    true,
    3,
    "a87ff679a2f3e71d9181a67b7542122c",
    { e4da3b7fbbce2345d7772b0674a318d5: 5 },
    /6/g,
    /0/g,
    function(){},
    "8f14e45fceea167a5a36dedd4bea2543",
    4,
    8,
    "1679091c5a880faf6fb5e6087eb1b2dc",
    true,
    [5, "e4da3b7fbbce2345d7772b0674a318d5"]
  ];

// Array 2
[
    false,
    3,
    7,
    8,
    function() {},
    /1/g,
    [6, "1679091c5a880faf6fb5e6087eb1b2dc"],
    2,
    7,
    function() {},
    [8, "c9f0f895fb98ab9159f51fd0297e236d"],
    function() {},
    5,
    1,
    "cfcd208495d565ef66e7dff9f98764da",
    /6/g
  ]

//  Union of array 1 and array 2 (your script)
[
    { eccbc87e4b5ce2fe28308fd9f2a7baf3: 3 },
    6,
    3,
    true,
    "a87ff679a2f3e71d9181a67b7542122c",
    { e4da3b7fbbce2345d7772b0674a318d5: 5 },
    /6/g,
    /0/g,
    function() {},
    "8f14e45fceea167a5a36dedd4bea2543",
    4,
    8,
    "1679091c5a880faf6fb5e6087eb1b2dc",
    [5, "e4da3b7fbbce2345d7772b0674a318d5"],
    false,
    7,
    function() {},
    /1/g,
    [6, "1679091c5a880faf6fb5e6087eb1b2dc"],
    2,
    [8, "c9f0f895fb98ab9159f51fd0297e236d"],
    5,
    1,
    "cfcd208495d565ef66e7dff9f98764da",
  ]

// Union of array 1 and array 2 (Sugar)
[
    { eccbc87e4b5ce2fe28308fd9f2a7baf3: 3 },
    6,
    3,
    true,
    "a87ff679a2f3e71d9181a67b7542122c",
    { e4da3b7fbbce2345d7772b0674a318d5: 5},
    /0/g,
    function() {},
    "8f14e45fceea167a5a36dedd4bea2543",
    4,
    8,
    "1679091c5a880faf6fb5e6087eb1b2dc",
    [5, "e4da3b7fbbce2345d7772b0674a318d5"],
    false,
    7,
    function() {},
    2,
    function() {},
    "c9f0f895fb98ab9159f51fd0297e236d",
    function() {},
    5,
    1,
    "cfcd208495d565ef66e7dff9f98764da",
  ]
```

First, as you pointed out there are a couple of errors in the Sugar version.
1. `[8, "c9f0f895fb98ab9159f51fd0297e236d"]` this element was found incorrectly in the resulting array separately, ie. it was brought into the result as separate elements instead of a nested array as it properly should have been.
2. Regexes found in the original arrays are not in the resulting array.

However, your script also had an error:
1. Functions found in the original array are not in the resulting array.

`1` was due to a bug where `Array#union` was incorrectly flattening arrays being passed to it. I have added a unit test for this and fixed the problem.

`2` was due to another, deeper issue with `Array#unique`, which I won't go into here but had farther reaching implications that could effect other methods by incorrectly interpreting regexes and functions found in arrays instead of performing a direct `equal` operation on them. It was a very good opportunity for cleanup, so I've added some tests here and fixed this as well.

This will fix any errors in these methods and will go out in the next release.

Now on to the biggest issue, which is the performance. Over twitter I had said that "this won't work for objects in arrays"... by that what I meant is that "this won't work for objects within arrays without some kind of an object hashing algorithm". And, in fact that is exactly what you have done for your test.

Now, I am aware of these methods of alleviating issues for O(n²) problems by leveraging hashes. In fact, `Array#union` and `Array#intersect` aren't the only ones that could benefit from this but in fact `Array#unique` most especially and others as well. However the problem is the cost involved.

First, this technique requires hashing objects based on their content, which does involve a performance cost. Whether or not that performance cost would outweigh the benefit gained is a question that can only be answered by benchmarking not only large arrays, but also various combinations of large objects and large arrays. I would argue that having large and complex JSON data structures is a much more common use-case than having massive arrays (at least on the scale of 10s of thousands and higher) and here is where I think the cost would start to come into play.

However, there is a bigger problem and that is the cost of introducing an external hashing algorithm, which is exactly what you've done in your example. I'm mostly worried about the added bulk to the codebase but also the problem of an external dependency that would have to be proven to work right. Lastly I would also be worried about this dependency should I decide to split up the library further, any methods that require `unique` would also require this dependency as well.

Finally there is the issue of hashing functions, which should never be considered "equal" unless they are equal by reference, and this is where the errors in your script are encountered.

So in summary, if all things were equal I would love to be able to do O(n log n) and gain that much more speed on array methods like this, but I don't think it should be done by adding a new large dependency, one that may actually hinder performance on large objects, and has a clear weakness when considering functions as well.

I'm interested in more feedback and if there's another angle or another way in which this could be implemented, I'm open to it.

In the meantime your test has already improved Sugar in the errors you've uncovered so thank you much for that!
